### PR TITLE
Added a takeUntil stage, which is an inclusive version of takeWhile

### PR DESCRIPTION
**Pull request by Falmarri**
This might be too specific, but I use this a lot in my current project and thought it might be useful upstream

### CODE REVIEW

**Comment by johanandren** [comment]
OK TO TEST

**Comment by johanandren** [comment]
Needs mention in `stages-overview.rst` as well.

**Comment by johanandren** [comment]
LGTM after `stages-overview.rst` updates

**Comment by johanandren** [comment]
MiMa check failed, it is safe however to whitelist as `FlowOps` isn't public to subclass/implement outside of Akka Streams. Whitelisting is done by adding the filters in `project/MiMa.scala`

```
[info] akka-stream: found 1 potential binary incompatibilities while checking against com.typesafe.akka:akka-stream_2.11:2.4.2  (filtered 135)
[error]  * method takeUntil(scala.Function1)akka.stream.scaladsl.FlowOps in trait akka.stream.scaladsl.FlowOps is present only in current version
[error]    filter with: ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.stream.scaladsl.FlowOps.takeUntil")
```

File: akka-docs/rst/scala/stream/stages-overview.rst [hunk 1]
```diff
@@ -587,6 +587,17 @@ when the predicate first return false and then complete.
...
+^^^^^^^^^
+Pass elements downstream until a predicate function returns false, and then emit that element and complete. This
+differs from takeWhile in that it's inclusive of the final element that causes the predicate to reutrn false
+
+**emits** while the predicate is true and until the first false result
+
+**backpressures** when downstream backpressures
+
+**completes** when predicate returned false, after emitting that final element, or upstream completes
+
```
Comment (johanandren) [inline akka-docs/rst/scala/stream/stages-overview.rst]: Great but there is another one where you need to add it: `akka-docs/rst/java/stream/stages-overview.rst`

File: akka-docs/rst/scala/stream/stages-overview.rst [hunk 2]
```diff
@@ -587,6 +587,17 @@ when the predicate first return false and then complete.
 
 **completes** when predicate returned false or upstream completes
 
+takeUntil
```
Comment (patriknw) [inline akka-docs/rst/scala/stream/stages-overview.rst]: isn't the name confusing? compare with scala range `(1 until 5)`

If we really need this operator, should it be a parameter to `takeWhile`?

Comment (Falmarri) [inline akka-docs/rst/scala/stream/stages-overview.rst]: You're right. This is confusing... Do you think it's worthwhile to parameterize takeWhile (that will break binary compatibility, right)? I still thing have the option to be inclusive of the predicate is useful, but if I'm just cluttering up the code I can hold off until the next major version?

Comment (patriknw) [inline akka-docs/rst/scala/stream/stages-overview.rst]: I dont think adding an overloaded method will break bin compat

Comment (patriknw) [inline akka-docs/rst/scala/stream/stages-overview.rst]: Not very efficient, but it's possible to express this with mapConcat and ordinary takeWhile

File: akka-docs/rst/scala/stream/stages-overview.rst (hunk 1 above)
Comment (ktoso) [inline akka-docs/rst/scala/stream/stages-overview.rst]: Can we write that test that makes sure stages are documented now? ;-)

**Comment by drewhk** [comment]
I opt for adding the extra flag, that can be done in bincompat way.

**Comment by drewhk** [comment]
The logic is LGTM

**Comment by johanandren** [comment]
@Falmarri any plans to complete this or should we either close or take it over?

**Comment by ktoso** [comment]
ping @Falmarri, would you be able to push this over the finish line?

**Comment by Falmarri** [comment]
Sorry. I kinda totally forgot about this and haven't come back to it. I'll take another look at what's left and get a new patch up

**Comment by ktoso** [comment]
Thanks a lot, would be awesome :-)
Let me know if help is needed :)

**Comment by Falmarri** [comment]
@ktoso is your jenkins server down? I can't see the details of the test failure

**Comment by johanandren** [comment]
Yeah, it crashed. Back up now though.

**Comment by monktastic** [comment]
Yay, +1 for this. I've implemented it in my codebase as well. Two example use cases:

1. Given a monotonically increasing, potentially non-contiguous stream (e.g., sequence numbers from eventsByPersistenceId), terminate with the greatest not exceeding n (and terminating as soon as n is seen, instead of waiting).
`def stopAt[T](f: T => Int)(n: Int) = Flow[T].takeWhile(f(_) <= n) via takeUntil(f(_) >= n)`

2. Using Flow.scan[Xor...] and short-circuiting so that the first Left is returned, if any.

File: akka-stream/src/main/scala/akka/stream/javadsl/SubFlow.scala [hunk 3]
```diff
@@ -561,7 +561,7 @@ class SubFlow[-In, +Out, +Mat](delegate: scaladsl.SubFlow[Out, Mat, scaladsl.Flo
 
   /**
    * Terminate processing (and cancel the upstream publisher) after predicate
-   * returns false for the first time. Due to input buffering some elements may have been
+   * returns false for the first time (or 1 after if inclusive). Due to input buffering some elements may have been
```
Comment (monktastic) [inline akka-stream/src/main/scala/akka/stream/javadsl/SubFlow.scala]: Suggestion:
`* returns false for the first time, including the first failed element iff inclusive is true.`

**Comment by patriknw** [comment]
@Falmarri This looks like it's pretty much ready, but take a look at the CI build log

I'm not sure what it is. Noticed this: https://jenkins.akka.io:8498/job/pr-validator-per-commit-jenkins/7099/
```
[error] /localhome/jenkinsakka/workspace/pr-validator-per-commit-jenkins/akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala:91: unclosed comment
[error] /**
[error] ^
[error] one error found
```

**Comment by Falmarri** [comment]
Looks like I screwed up the rebase. Should be good now.

**Comment by Falmarri** [comment]
Is this failure my issue? Or something unrelated?

**Comment by johanandren** [comment]
These are the warnings, from our binary compatibility checking tool MiMa, all those interfaces are for internal implementation/subclassing only so it should be safe to add the filter excludes described in the output:

```
[info] akka-stream: found 7 potential binary incompatibilities while checking against com.typesafe.akka:akka-stream_2.11:2.4.2  (filtered 279)
[error]  * method takeWhile(akka.japi.function.Predicate)akka.stream.javadsl.SubSource in class akka.stream.javadsl.SubSource does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("akka.stream.javadsl.SubSource.takeWhile")
[error]  * method takeWhile(akka.japi.function.Predicate)akka.stream.javadsl.SubFlow in class akka.stream.javadsl.SubFlow does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("akka.stream.javadsl.SubFlow.takeWhile")
[error]  * method takeWhile(scala.Function1)akka.stream.scaladsl.FlowOps in class akka.stream.scaladsl.Flow does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("akka.stream.scaladsl.Flow.takeWhile")
[error]  * method takeWhile(scala.Function1)akka.stream.scaladsl.FlowOps in class akka.stream.scaladsl.Source does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("akka.stream.scaladsl.Source.takeWhile")
[error]  * method takeWhile(scala.Function1)akka.stream.scaladsl.FlowOps in trait akka.stream.scaladsl.FlowOps does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("akka.stream.scaladsl.FlowOps.takeWhile")
[error]  * method takeWhile(scala.Function1,Boolean)akka.stream.scaladsl.FlowOps in trait akka.stream.scaladsl.FlowOps is present only in current version
[error]    filter with: ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.stream.scaladsl.FlowOps.takeWhile")
[error]  * synthetic method takeWhile$default$2()Boolean in trait akka.stream.scaladsl.FlowOps is present only in current version
[error]    filter with: ProblemFilters.exclude[ReversedMissingMethodProblem]
```

**Comment by Falmarri** [comment]
@johanandren hmm. thinking about that now, are you sure that's binary compatible? Particularly the scaladsl.Flow and scaladsl.Source?

**Comment by johanandren** [comment]
Ah, no you are right, interpreted it as the ticket still was about adding a new factory method, but that has changed, and now it is about adding a parameter for takeWhile. The old method needs to stay in all places, and the boolean needs to be added alongside it, when I quickly skimmed the pr it seems as though you mostly did that but not everywhere?

**Comment by Falmarri** [comment]
@johanandren I did it for java, but added the argument for scala. I'll add a second scala method with a TODO to combine them when you guys move to a backwards incompatible version

**Comment by johanandren** [comment]
As there is no ticket to track this I thought we can use this new ticket #21330?

**Comment by patriknw** [comment]
seems like almost nothing left here, I'll do the final adjustments

**Comment by patriknw** [comment]
Completed in  https://github.com/akka/akka/pull/21687

**Pull request closed without merging**